src/openassetio-core/include/openassetio/trait/TraitBase.hpp

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2013-2022 The Foundry Visionmongers Ltd
+/**
+ * Base class for all specification traits.
+ */
+#pragma once
+#include <algorithm>
+#include <memory>
+#include <utility>
+
+#include "../specification/Specification.hpp"
+
+namespace openassetio {
+inline namespace OPENASSETIO_VERSION {
+namespace trait {
+
+/**
+ * Abstract CRTP base class for specification traits.
+ *
+ * @tparam Derived Concrete subclass.
+ */
+template <class Derived>
+struct TraitBase {
+  /// Ref-counted smart pointer to underlying Specification.
+  using SpecificationPtr = std::shared_ptr<specification::Specification>;
+
+  /**
+   * Construct this trait, wrapping given specification.
+   *
+   * @param specification Underlying specification to wrap.
+   */
+  explicit TraitBase(SpecificationPtr specification) : specification_{std::move(specification)} {}
+
+  /**
+   * Get the ID of this trait.
+   *
+   * @return ID of this trait.
+   */
+  [[nodiscard]] static const TraitId& id() { return Derived::kId; }
```

Comment:
[fix] Capturing an off-line conversation:

If we don't need to pass around base class `TraitPtr`s, then do we need this method? Is `kId` sufficient?